File: src/models/LevelCatalog.ts
import { Direction, DIRECTION_OFFSETS, IllustrationType, LevelData } from '../types';
import { SeededRandom } from './prng';

export interface WordDefinition {
  word: string;
  icon: string;
  hintText: string;
  illustrationType: IllustrationType;
}

export const LEVEL_WORDS: WordDefinition[] = [
  // 1 - 10: Original Classics
  { word: "APPLE", icon: "🍎", hintText: "A sweet, crunchy red fruit!", illustrationType: IllustrationType.APPLE },
  { word: "CAT", icon: "🐱", hintText: "A cute furry friend that says meow!", illustrationType: IllustrationType.CAT },
  { word: "FLOWER", icon: "🌸", hintText: "Blooms in spring with colourful petals!", illustrationType: IllustrationType.FLOWER },
  { word: "SUN", icon: "☀️", hintText: "Shines bright and warm in the sky!", illustrationType: IllustrationType.SUN },
  { word: "FISH", icon: "🐟", hintText: "Swims gracefully under the blue sea!", illustrationType: IllustrationType.FISH },
  { word: "RAINBOW", icon: "🌈", hintText: "Magical colours after the rain!", illustrationType: IllustrationType.RAINBOW },
  { word: "BALLOON", icon: "🎈", hintText: "Floats high in the air at parties!", illustrationType: IllustrationType.BALLOON },
  { word: "BUTTERFLY", icon: "🦋", hintText: "Flutters gently with painted wings!", illustrationType: IllustrationType.BUTTERFLY },
  { word: "AVOCADO", icon: "🥑", hintText: "Creamy green fruit with a round seed!", illustrationType: IllustrationType.AVOCADO },
  { word: "ROCKET", icon: "🚀", hintText: "Zooms into outer space to the stars!", illustrationType: IllustrationType.ROCKET },

  // 11 - 20: Animals & Pets
  { word: "DOG", icon: "🐶", hintText: "Loyal furry best friend that barks!", illustrationType: IllustrationType.CAT },
  { word: "BIRD", icon: "🐦", hintText: "Sings sweet morning songs on trees!", illustrationType: IllustrationType.BUTTERFLY },
  { word: "LION", icon: "🦁", hintText: "The brave king of the golden savanna!", illustrationType: IllustrationType.CAT },
  { word: "PANDA", icon: "🐼", hintText: "Gentle bear that loves chewing bamboo!", illustrationType: IllustrationType.CAT },
  { word: "BEE", icon: "🐝", hintText: "Buzzes near blooms collecting honey!", illustrationType: IllustrationType.FLOWER },
  { word: "DUCK", icon: "🦆", hintText: "Quacks happily paddling on the pond!", illustrationType: IllustrationType.FISH },
  { word: "FROG", icon: "🐸", hintText: "Hops over green lily pads in the lake!", illustrationType: IllustrationType.FISH },
  { word: "TIGER", icon: "🐯", hintText: "Strong big cat with bold orange stripes!", illustrationType: IllustrationType.CAT },
  { word: "HORSE", icon: "🐴", hintText: "Gallops fast across lush open fields!", illustrationType: IllustrationType.CAT },
  { word: "WHALE", icon: "🐋", hintText: "Majestic giant swimming deep in oceans!", illustrationType: IllustrationType.FISH },

  // 21 - 30: Nature & Sky
  { word: "TREE", icon: "🌳", hintText: "Tall leafy canopy providing cool shade!", illustrationType: IllustrationType.FLOWER },
  { word: "MOON", icon: "🌙", hintText: "Glows gently in the starry night sky!", illustrationType: IllustrationType.SUN },
  { word: "STAR", icon: "⭐", hintText: "Twinkles brightly from millions of miles!", illustrationType: IllustrationType.SUN },
  { word: "CLOUD", icon: "☁️", hintText: "Soft puffy white cotton ball in the sky!", illustrationType: IllustrationType.RAINBOW },
  { word: "LEAF", icon: "🍃", hintText: "Dances in the crisp autumn breeze!", illustrationType: IllustrationType.FLOWER },
  { word: "RIVER", icon: "🌊", hintText: "Clear sparkling stream flowing downhill!", illustrationType: IllustrationType.FISH },
  { word: "RAIN", icon: "🌧️", hintText: "Pattering drops watering thirsty garden soil!", illustrationType: IllustrationType.RAINBOW },
  { word: "SNOW", icon: "❄️", hintText: "Delicate icy flakes creating winter magic!", illustrationType: IllustrationType.RAINBOW },
  { word: "WIND", icon: "💨", hintText: "Invisible breeze that turns paper pinwheels!", illustrationType: IllustrationType.BALLOON },
  { word: "BEACH", icon: "🏖️", hintText: "Golden sand warm under the summer sun!", illustrationType: IllustrationType.SUN },

  // 31 - 40: Fruits & Berries
  { word: "BANANA", icon: "🍌", hintText: "Sweet yellow energy treat monkeys love!", illustrationType: IllustrationType.APPLE },
  { word: "CHERRY", icon: "🍒", hintText: "Pair of glossy red jewels on a stem!", illustrationType: IllustrationType.APPLE },
  { word: "ORANGE", icon: "🍊", hintText: "Juicy citrus slice bursting with vitamin C!", illustrationType: IllustrationType.APPLE },
  { word: "GRAPE", icon: "🍇", hintText: "Plump sweet cluster growing on vineyard vines!", illustrationType: IllustrationType.APPLE },
  { word: "LEMON", icon: "🍋", hintText: "Zesty yellow fruit that gives lemonade kick!", illustrationType: IllustrationType.APPLE },
  { word: "PEACH", icon: "🍑", hintText: "Velvety fuzzy sweet summer delicacy!", illustrationType: IllustrationType.APPLE },
  { word: "PEAR", icon: "🍐", hintText: "Bell-shaped juicy fruit with golden skin!", illustrationType: IllustrationType.APPLE },
  { word: "BERRY", icon: "🍓", hintText: "Tiny sweet fruit packed with woodland flavour!", illustrationType: IllustrationType.APPLE },
  { word: "MELON", icon: "🍈", hintText: "Refreshing cool melon on a picnic blanket!", illustrationType: IllustrationType.AVOCADO },
  { word: "MANGO", icon: "🥭", hintText: "Tropical king of sweetness and aroma!", illustrationType: IllustrationType.AVOCADO },

  // 41 - 50: Fun Toys & Celebration
  { word: "KITE", icon: "🪁", hintText: "Soars and loops high against blue clouds!", illustrationType: IllustrationType.BALLOON },
  { word: "DRUM", icon: "🥁", hintText: "Bangs the joyful rhythm of school bands!", illustrationType: IllustrationType.BALLOON },
  { word: "BELL", icon: "🔔", hintText: "Rings clearly to gather everyone together!", illustrationType: IllustrationType.FLOWER },
  { word: "GIFT", icon: "🎁", hintText: "Wrapped in colorful paper with a big ribbon!", illustrationType: IllustrationType.BALLOON },
  { word: "BALL", icon: "⚽", hintText: "Bounces across playgrounds and sunny courts!", illustrationType: IllustrationType.BALLOON },
  { word: "CAKE", icon: "🎂", hintText: "Sweet celebration dessert with birthday candles!", illustrationType: IllustrationType.BALLOON },
  { word: "COOKIE", icon: "🍪", hintText: "Baked treat loaded with chocolate chips!", illustrationType: IllustrationType.APPLE },
  { word: "CANDY", icon: "🍬", hintText: "Delicious sugary wrapper full of sweetness!", illustrationType: IllustrationType.BALLOON },
  { word: "PIZZA", icon: "🍕", hintText: "Cheesy slice fresh from the stone oven!", illustrationType: IllustrationType.AVOCADO },
  { word: "JUICE", icon: "🧃", hintText: "Chilled fruity drink with a paper straw!", illustrationType: IllustrationType.APPLE },

  // 51 - 60: Ocean & Waters
  { word: "SHARK", icon: "🦈", hintText: "Swift ocean swimmer with a dorsal fin!", illustrationType: IllustrationType.FISH },
  { word: "CRAB", icon: "🦀", hintText: "Walks sideways with clicking red pincers!", illustrationType: IllustrationType.FISH },
  { word: "CORAL", icon: "🪸", hintText: "Vibrant underwater home for playful fish!", illustrationType: IllustrationType.FISH },
  { word: "SHELL", icon: "🐚", hintText: "Treasured ocean souvenir found on sandy shore!", illustrationType: IllustrationType.FISH },
  { word: "BOAT", icon: "⛵", hintText: "Sails smoothly across peaceful rippling bays!", illustrationType: IllustrationType.FISH },
  { word: "WAVE", icon: "🌊", hintText: "Curls and crashes into sparkling white foam!", illustrationType: IllustrationType.FISH },
  { word: "TURTLE", icon: "🐢", hintText: "Glides gracefully through coral sea reefs!", illustrationType: IllustrationType.FISH },
  { word: "ISLAND", icon: "🏝️", hintText: "Tropical haven shaded by coconut palms!", illustrationType: IllustrationType.SUN },
  { word: "DOLPHIN", icon: "🐬", hintText: "Clever friendly sea acrobat leaping waves!", illustrationType: IllustrationType.FISH },
  { word: "OCTOPUS", icon: "🐙", hintText: "Eight clever tentacles exploring tide pools!", illustrationType: IllustrationType.FISH },

  // 61 - 70: Cosmos & Exploration
  { word: "PLANET", icon: "🪐", hintText: "Giant world spinning around the sun!", illustrationType: IllustrationType.ROCKET },
  { word: "COMET", icon: "☄️", hintText: "Cosmic snowball blazing a glowing tail!", illustrationType: IllustrationType.ROCKET },
  { word: "EARTH", icon: "🌍", hintText: "Our beautiful blue home in the universe!", illustrationType: IllustrationType.ROCKET },
  { word: "MARS", icon: "🔴", hintText: "The intriguing red planet in starry night!", illustrationType: IllustrationType.ROCKET },
  { word: "ORBIT", icon: "🛰️", hintText: "Looping path celestial objects travel on!", illustrationType: IllustrationType.ROCKET },
  { word: "GALAXY", icon: "🌌", hintText: "Spiralling ocean of billions of bright stars!", illustrationType: IllustrationType.ROCKET },
  { word: "METEOR", icon: "🌠", hintText: "Shooting light making wishes come true!", illustrationType: IllustrationType.ROCKET },
  { word: "ALIEN", icon: "👽", hintText: "Friendly cosmic visitor from far constellations!", illustrationType: IllustrationType.ROCKET },
  { word: "TELESCOPE", icon: "🔭", hintText: "Peeks at distant rings of Saturn!", illustrationType: IllustrationType.ROCKET },
  { word: "SPACE", icon: "🛸", hintText: "Infinite wondrous playground of the cosmos!", illustrationType: IllustrationType.ROCKET },

  // 71 - 80: Garden & Blooms
  { word: "ROSE", icon: "🌹", hintText: "Fragrant flower with velvety deep petals!", illustrationType: IllustrationType.FLOWER },
  { word: "TULIP", icon: "🌷", hintText: "Cup-shaped spring bloom upright in flowerbeds!", illustrationType: IllustrationType.FLOWER },
  { word: "SEED", icon: "🌱", hintText: "Sprouts green new life in healthy soil!", illustrationType: IllustrationType.FLOWER },
  { word: "DAISY", icon: "🌼", hintText: "Cheerful white petals with a golden center!", illustrationType: IllustrationType.FLOWER },
  { word: "GARDEN", icon: "🪴", hintText: "Quiet sanctuary filled with herbs and blooms!", illustrationType: IllustrationType.FLOWER },
  { word: "MUSHROOM", icon: "🍄", hintText: "Cute forest umbrella sheltering beetles!", illustrationType: IllustrationType.FLOWER },
  { word: "GRASS", icon: "🌾", hintText: "Soft green carpet spreading over hills!", illustrationType: IllustrationType.FLOWER },
  { word: "BEETLE", icon: "🐞", hintText: "Little red ladybug with spotted wing covers!", illustrationType: IllustrationType.BUTTERFLY },
  { word: "SPIDER", icon: "🕷️", hintText: "Weaves intricate geometric dewy webs!", illustrationType: IllustrationType.BUTTERFLY },
  { word: "CATERPILLAR", icon: "🐛", hintText: "Munches green leaves before becoming butterfly!", illustrationType: IllustrationType.BUTTERFLY },

  // 81 - 90: Learning & Creativity
  { word: "BOOK", icon: "📚", hintText: "Portal to thrilling tales and grand adventures!", illustrationType: IllustrationType.BALLOON },
  { word: "PENCIL", icon: "✏️", hintText: "Sketches dreams and solves puzzle grids!", illustrationType: IllustrationType.APPLE },
  { word: "CRAYON", icon: "🖍️", hintText: "Adds vibrant rainbow colors to sketches!", illustrationType: IllustrationType.RAINBOW },
  { word: "PAINT", icon: "🎨", hintText: "Swirls vivid hues across blank canvases!", illustrationType: IllustrationType.RAINBOW },
  { word: "PAPER", icon: "📄", hintText: "Folds into airplanes and origami swans!", illustrationType: IllustrationType.BALLOON },
  { word: "CLOCK", icon: "⏰", hintText: "Ticks happily counting cheerful moments!", illustrationType: IllustrationType.SUN },
  { word: "MUSIC", icon: "🎵", hintText: "Sweet harmonious notes that inspire dancing!", illustrationType: IllustrationType.BALLOON },
  { word: "PUZZLE", icon: "🧩", hintText: "Fitting clever pieces together with joy!", illustrationType: IllustrationType.BALLOON },
  { word: "SMILE", icon: "😊", hintText: "Bright friendly curve that brightens anyone's day!", illustrationType: IllustrationType.SUN },
  { word: "HEART", icon: "❤️", hintText: "Symbol of warmth, friendship and care!", illustrationType: IllustrationType.APPLE },

  // 91 - 100: Grand Adventures & Outdoors
  { word: "CAMP", icon: "⛺", hintText: "Cozy tent pitched beside starlit campfires!", illustrationType: IllustrationType.SUN },
  { word: "MOUNTAIN", icon: "🏔️", hintText: "Majestic snowy peak reaching towards clouds!", illustrationType: IllustrationType.SUN },
  { word: "FOREST", icon: "🌲", hintText: "Whispering evergreen woods full of wonders!", illustrationType: IllustrationType.FLOWER },
  { word: "SUNSET", icon: "🌅", hintText: "Golden evening horizon glowing purple and pink!", illustrationType: IllustrationType.SUN },
  { word: "CASTLE", icon: "🏰", hintText: "Enchanted stone fortress from fairy tale books!", illustrationType: IllustrationType.BALLOON },
  { word: "TREASURE", icon: "💎", hintText: "Glittering hidden chest of sparkling crystals!", illustrationType: IllustrationType.SUN },
  { word: "HERO", icon: "🦸", hintText: "Brave champion who helps friends every day!", illustrationType: IllustrationType.ROCKET },
  { word: "WIZARD", icon: "🧙", hintText: "Sparks magic wand with glittering constellations!", illustrationType: IllustrationType.ROCKET },
  { word: "CROWN", icon: "👑", hintText: "Golden regal tiara adorned with rubies!", illustrationType: IllustrationType.SUN },
  { word: "CHAMPION", icon: "🏆", hintText: "Proud golden trophy for completing 100 levels!", illustrationType: IllustrationType.SUN }
];

export function computeTargetCells(
  startRow: number,
  startCol: number,
  wordLength: number,
  direction: Direction
): [number, number][] {
  const offset = DIRECTION_OFFSETS[direction];
  const cells: [number, number][] = [];
  for (let i = 0; i < wordLength; i++) {
    cells.push([startRow + i * offset.dRow, startCol + i * offset.dCol]);
  }
  return cells;
}

export function createLevel(levelNum: number): LevelData {
  const wordIndex = Math.min(Math.max(0, levelNum - 1), LEVEL_WORDS.length - 1);
  const def = LEVEL_WORDS[wordIndex];
  const word = def.word;

  let baseDim: number;
  if (levelNum <= 10) {
    baseDim = word.length + 1;
  } else if (levelNum <= 40) {
    baseDim = Math.max(word.length + 1, 6);
  } else if (levelNum <= 75) {
    baseDim = Math.max(word.length + 1, 7);
  } else {
    baseDim = Math.max(word.length + 1, 8);
  }

  const gridRows = Math.max(baseDim, word.length);
  const gridCols = Math.max(baseDim, word.length);

  const candidateDirections: Direction[] =
    levelNum <= 5
      ? [Direction.HORIZONTAL, Direction.VERTICAL]
      : [Direction.HORIZONTAL, Direction.VERTICAL, Direction.DIAGONAL_DOWN, Direction.DIAGONAL_UP];

  const rng = new SeededRandom(levelNum * 997 + 13);
  const shuffledDirs = rng.shuffle(candidateDirections);

  let chosenDirection = Direction.HORIZONTAL;
  let chosenStartRow = 0;
  let chosenStartCol = 0;
  let placed = false;

  for (const dir of shuffledDirs) {
    const offset = DIRECTION_OFFSETS[dir];
    const minStartRow = offset.dRow < 0 ? word.length - 1 : 0;
    const maxStartRow = offset.dRow > 0 ? gridRows - word.length : gridRows - 1;

    const minStartCol = offset.dCol < 0 ? word.length - 1 : 0;
    const maxStartCol = offset.dCol > 0 ? gridCols - word.length : gridCols - 1;

    if (maxStartRow >= minStartRow && maxStartCol >= minStartCol) {
      const rowRange = maxStartRow - minStartRow + 1;
      const colRange = maxStartCol - minStartCol + 1;
      chosenStartRow = minStartRow + rng.nextInt(rowRange);
      chosenStartCol = minStartCol + rng.nextInt(colRange);
      chosenDirection = dir;
      placed = true;
      break;
    }
  }

  if (!placed) {
    chosenDirection = Direction.HORIZONTAL;
    chosenStartRow = Math.min(Math.floor(gridRows / 2), gridRows - 1);
    chosenStartCol = 0;
  }

  const chapterIndex = Math.min(Math.floor((levelNum - 1) / 10), 9);
  const categoryNames = [
    'Sunny Meadow',
    'Safari & Pets',
    'Whispering Woods',
    'Sweet Orchard',
    'Toy Fair & Sweets',
    'Coral Waters',
    'Cosmic Sky',
    'Blossom Garden',
    'Cozy Village',
    'Grand Discovery',
  ];

  return {
    levelNumber: levelNum,
    word,
    icon: def.icon,
    hintText: def.hintText,
    category: categoryNames[chapterIndex],
    gridRows,
    gridCols,
    startRow: chosenStartRow,
    startCol: chosenStartCol,
    direction: chosenDirection,
    illustrationType: def.illustrationType,
    targetCells: computeTargetCells(chosenStartRow, chosenStartCol, word.length, chosenDirection),
  };
}

export const ALL_100_LEVELS: LevelData[] = Array.from({ length: 100 }, (_, i) => createLevel(i + 1));

File: app/src/main/java/com/example/model/LevelCatalog.kt
package com.example.model

/**
 * 100 distinct word definitions tailored for casual, colourful Word Search puzzles.
 * Grouped across cheerful kid-friendly themes: Animals, Nature, Fruits & Veggies,
 * Objects, Space, Ocean, Sweets, School, Colors, and Everyday Wonders.
 */
data class WordDefinition(
    val word: String,
    val icon: String,
    val hintText: String,
    val illustrationType: IllustrationType
)

object LevelCatalog {

    val WORDS: List<WordDefinition> = listOf(
        // 1 - 10: Original Classics
        WordDefinition("APPLE", "🍎", "A sweet, crunchy red fruit!", IllustrationType.APPLE),
        WordDefinition("CAT", "🐱", "A cute furry friend that says meow!", IllustrationType.CAT),
        WordDefinition("FLOWER", "🌸", "Blooms in spring with colourful petals!", IllustrationType.FLOWER),
        WordDefinition("SUN", "☀️", "Shines bright and warm in the sky!", IllustrationType.SUN),
        WordDefinition("FISH", "🐟", "Swims gracefully under the blue sea!", IllustrationType.FISH),
        WordDefinition("RAINBOW", "🌈", "Magical colours after the rain!", IllustrationType.RAINBOW),
        WordDefinition("BALLOON", "🎈", "Floats high in the air at parties!", IllustrationType.BALLOON),
        WordDefinition("BUTTERFLY", "🦋", "Flutters gently with painted wings!", IllustrationType.BUTTERFLY),
        WordDefinition("AVOCADO", "🥑", "Creamy green fruit with a round seed!", IllustrationType.AVOCADO),
        WordDefinition("ROCKET", "🚀", "Zooms into outer space to the stars!", IllustrationType.ROCKET),

        // 11 - 20: Animals & Pets
        WordDefinition("DOG", "🐶", "Loyal furry best friend that barks!", IllustrationType.CAT),
        WordDefinition("BIRD", "🐦", "Sings sweet morning songs on trees!", IllustrationType.BUTTERFLY),
        WordDefinition("LION", "🦁", "The brave king of the golden savanna!", IllustrationType.CAT),
        WordDefinition("PANDA", "🐼", "Gentle bear that loves chewing bamboo!", IllustrationType.CAT),
        WordDefinition("BEE", "🐝", "Buzzes near blooms collecting honey!", IllustrationType.FLOWER),
        WordDefinition("DUCK", "🦆", "Quacks happily paddling on the pond!", IllustrationType.FISH),
        WordDefinition("FROG", "🐸", "Hops over green lily pads in the lake!", IllustrationType.FISH),
        WordDefinition("TIGER", "🐯", "Strong big cat with bold orange stripes!", IllustrationType.CAT),
        WordDefinition("HORSE", "🐴", "Gallops fast across lush open fields!", IllustrationType.CAT),
        WordDefinition("WHALE", "🐋", "Majestic giant swimming deep in oceans!", IllustrationType.FISH),

        // 21 - 30: Nature & Sky
        WordDefinition("TREE", "🌳", "Tall leafy canopy providing cool shade!", IllustrationType.FLOWER),
        WordDefinition("MOON", "🌙", "Glows gently in the starry night sky!", IllustrationType.SUN),
        WordDefinition("STAR", "⭐", "Twinkles brightly from millions of miles!", IllustrationType.SUN),
        WordDefinition("CLOUD", "☁️", "Soft puffy white cotton ball in the sky!", IllustrationType.RAINBOW),
        WordDefinition("LEAF", "🍃", "Dances in the crisp autumn breeze!", IllustrationType.FLOWER),
        WordDefinition("RIVER", "🌊", "Clear sparkling stream flowing downhill!", IllustrationType.FISH),
        WordDefinition("RAIN", "🌧️", "Pattering drops watering thirsty garden soil!", IllustrationType.RAINBOW),
        WordDefinition("SNOW", "❄️", "Delicate icy flakes creating winter magic!", IllustrationType.RAINBOW),
        WordDefinition("WIND", "💨", "Invisible breeze that turns paper pinwheels!", IllustrationType.BALLOON),
        WordDefinition("BEACH", "🏖️", "Golden sand warm under the summer sun!", IllustrationType.SUN),

        // 31 - 40: Fruits & Berries
        WordDefinition("BANANA", "🍌", "Sweet yellow energy treat monkeys love!", IllustrationType.APPLE),
        WordDefinition("CHERRY", "🍒", "Pair of glossy red jewels on a stem!", IllustrationType.APPLE),
        WordDefinition("ORANGE", "🍊", "Juicy citrus slice bursting with vitamin C!", IllustrationType.APPLE),
        WordDefinition("GRAPE", "🍇", "Plump sweet cluster growing on vineyard vines!", IllustrationType.APPLE),
        WordDefinition("LEMON", "🍋", "Zesty yellow fruit that gives lemonade kick!", IllustrationType.APPLE),
        WordDefinition("PEACH", "🍑", "Velvety fuzzy sweet summer delicacy!", IllustrationType.APPLE),
        WordDefinition("PEAR", "🍐", "Bell-shaped juicy fruit with golden skin!", IllustrationType.APPLE),
        WordDefinition("BERRY", "🍓", "Tiny sweet fruit packed with woodland flavour!", IllustrationType.APPLE),
        WordDefinition("MELON", "🍈", "Refreshing cool melon on a picnic blanket!", IllustrationType.AVOCADO),
        WordDefinition("MANGO", "🥭", "Tropical king of sweetness and aroma!", IllustrationType.AVOCADO),

        // 41 - 50: Fun Toys & Celebration
        WordDefinition("KITE", "🪁", "Soars and loops high against blue clouds!", IllustrationType.BALLOON),
        WordDefinition("DRUM", "🥁", "Bangs the joyful rhythm of school bands!", IllustrationType.BALLOON),
        WordDefinition("BELL", "🔔", "Rings clearly to gather everyone together!", IllustrationType.FLOWER),
        WordDefinition("GIFT", "🎁", "Wrapped in colorful paper with a big ribbon!", IllustrationType.BALLOON),
        WordDefinition("BALL", "⚽", "Bounces across playgrounds and sunny courts!", IllustrationType.BALLOON),
        WordDefinition("CAKE", "🎂", "Sweet celebration dessert with birthday candles!", IllustrationType.BALLOON),
        WordDefinition("COOKIE", "🍪", "Baked treat loaded with chocolate chips!", IllustrationType.APPLE),
        WordDefinition("CANDY", "🍬", "Delicious sugary wrapper full of sweetness!", IllustrationType.BALLOON),
        WordDefinition("PIZZA", "🍕", "Cheesy slice fresh from the stone oven!", IllustrationType.AVOCADO),
        WordDefinition("JUICE", "🧃", "Chilled fruity drink with a paper straw!", IllustrationType.APPLE),

        // 51 - 60: Ocean & Waters
        WordDefinition("SHARK", "🦈", "Swift ocean swimmer with a dorsal fin!", IllustrationType.FISH),
        WordDefinition("CRAB", "🦀", "Walks sideways with clicking red pincers!", IllustrationType.FISH),
        WordDefinition("CORAL", "🪸", "Vibrant underwater home for playful fish!", IllustrationType.FISH),
        WordDefinition("SHELL", "🐚", "Treasured ocean souvenir found on sandy shore!", IllustrationType.FISH),
        WordDefinition("BOAT", "⛵", "Sails smoothly across peaceful rippling bays!", IllustrationType.FISH),
        WordDefinition("WAVE", "🌊", "Curls and crashes into sparkling white foam!", IllustrationType.FISH),
        WordDefinition("TURTLE", "🐢", "Glides gracefully through coral sea reefs!", IllustrationType.FISH),
        WordDefinition("ISLAND", "🏝️", "Tropical haven shaded by coconut palms!", IllustrationType.SUN),
        WordDefinition("DOLPHIN", "🐬", "Clever friendly sea acrobat leaping waves!", IllustrationType.FISH),
        WordDefinition("OCTOPUS", "🐙", "Eight clever tentacles exploring tide pools!", IllustrationType.FISH),

        // 61 - 70: Cosmos & Exploration
        WordDefinition("PLANET", "🪐", "Giant world spinning around the sun!", IllustrationType.ROCKET),
        WordDefinition("COMET", "☄️", "Cosmic snowball blazing a glowing tail!", IllustrationType.ROCKET),
        WordDefinition("EARTH", "🌍", "Our beautiful blue home in the universe!", IllustrationType.ROCKET),
        WordDefinition("MARS", "🔴", "The intriguing red planet in starry night!", IllustrationType.ROCKET),
        WordDefinition("ORBIT", "🛰️", "Looping path celestial objects travel on!", IllustrationType.ROCKET),
        WordDefinition("GALAXY", "🌌", "Spiralling ocean of billions of bright stars!", IllustrationType.ROCKET),
        WordDefinition("METEOR", "🌠", "Shooting light making wishes come true!", IllustrationType.ROCKET),
        WordDefinition("ALIEN", "👽", "Friendly cosmic visitor from far constellations!", IllustrationType.ROCKET),
        WordDefinition("TELESCOPE", "🔭", "Peeks at distant rings of Saturn!", IllustrationType.ROCKET),
        WordDefinition("SPACE", "🛸", "Infinite wondrous playground of the cosmos!", IllustrationType.ROCKET),

        // 71 - 80: Garden & Blooms
        WordDefinition("ROSE", "🌹", "Fragrant flower with velvety deep petals!", IllustrationType.FLOWER),
        WordDefinition("TULIP", "🌷", "Cup-shaped spring bloom upright in flowerbeds!", IllustrationType.FLOWER),
        WordDefinition("SEED", "🌱", "Sprouts green new life in healthy soil!", IllustrationType.FLOWER),
        WordDefinition("DAISY", "🌼", "Cheerful white petals with a golden center!", IllustrationType.FLOWER),
        WordDefinition("GARDEN", "🪴", "Quiet sanctuary filled with herbs and blooms!", IllustrationType.FLOWER),
        WordDefinition("MUSHROOM", "🍄", "Cute forest umbrella sheltering beetles!", IllustrationType.FLOWER),
        WordDefinition("GRASS", "🌾", "Soft green carpet spreading over hills!", IllustrationType.FLOWER),
        WordDefinition("BEETLE", "🐞", "Little red ladybug with spotted wing covers!", IllustrationType.BUTTERFLY),
        WordDefinition("SPIDER", "🕷️", "Weaves intricate geometric dewy webs!", IllustrationType.BUTTERFLY),
        WordDefinition("CATERPILLAR", "🐛", "Munches green leaves before becoming butterfly!", IllustrationType.BUTTERFLY),

        // 81 - 90: Learning & Creativity
        WordDefinition("BOOK", "📚", "Portal to thrilling tales and grand adventures!", IllustrationType.BALLOON),
        WordDefinition("PENCIL", "✏️", "Sketches dreams and solves puzzle grids!", IllustrationType.APPLE),
        WordDefinition("CRAYON", "🖍️", "Adds vibrant rainbow colors to sketches!", IllustrationType.RAINBOW),
        WordDefinition("PAINT", "🎨", "Swirls vivid hues across blank canvases!", IllustrationType.RAINBOW),
        WordDefinition("PAPER", "📄", "Folds into airplanes and origami swans!", IllustrationType.BALLOON),
        WordDefinition("CLOCK", "⏰", "Ticks happily counting cheerful moments!", IllustrationType.SUN),
        WordDefinition("MUSIC", "🎵", "Sweet harmonious notes that inspire dancing!", IllustrationType.BALLOON),
        WordDefinition("PUZZLE", "🧩", "Fitting clever pieces together with joy!", IllustrationType.BALLOON),
        WordDefinition("SMILE", "😊", "Bright friendly curve that brightens anyone's day!", IllustrationType.SUN),
        WordDefinition("HEART", "❤️", "Symbol of warmth, friendship and care!", IllustrationType.APPLE),

        // 91 - 100: Grand Adventures & Outdoors
        WordDefinition("CAMP", "⛺", "Cozy tent pitched beside starlit campfires!", IllustrationType.SUN),
        WordDefinition("MOUNTAIN", "🏔️", "Majestic snowy peak reaching towards clouds!", IllustrationType.SUN),
        WordDefinition("FOREST", "🌲", "Whispering evergreen woods full of wonders!", IllustrationType.FLOWER),
        WordDefinition("SUNSET", "🌅", "Golden evening horizon glowing purple and pink!", IllustrationType.SUN),
        WordDefinition("CASTLE", "🏰", "Enchanted stone fortress from fairy tale books!", IllustrationType.BALLOON),
        WordDefinition("TREASURE", "💎", "Glittering hidden chest of sparkling crystals!", IllustrationType.SUN),
        WordDefinition("HERO", "🦸", "Brave champion who helps friends every day!", IllustrationType.ROCKET),
        WordDefinition("WIZARD", "🧙", "Sparks magic wand with glittering constellations!", IllustrationType.ROCKET),
        WordDefinition("CROWN", "👑", "Golden regal tiara adorned with rubies!", IllustrationType.SUN),
        WordDefinition("CHAMPION", "🏆", "Proud golden trophy for completing 100 levels!", IllustrationType.SUN)
    )

    /**
     * Generates a fully verified, playable LevelData for the given level number (1..100).
     * Deterministic placement based on level index.
     */
    fun createLevel(levelNum: Int): LevelData {
        val wordIndex = (levelNum - 1).coerceIn(0, WORDS.size - 1)
        val def = WORDS[wordIndex]
        val word = def.word

        // Ensure grid is strictly large enough for the word length
        val baseDim = when {
            levelNum <= 10 -> word.length + 1
            levelNum <= 40 -> maxOf(word.length + 1, 6)
            levelNum <= 75 -> maxOf(word.length + 1, 7)
            else -> maxOf(word.length + 1, 8)
        }
        val gridRows = maxOf(baseDim, word.length)
        val gridCols = maxOf(baseDim, word.length)

        // Available directions based on level advancement:
        // Early levels: Horizontal & Vertical
        // Intermediate & Advanced: all 4 directions
        val candidateDirections = if (levelNum <= 5) {
            listOf(Direction.HORIZONTAL, Direction.VERTICAL)
        } else {
            listOf(Direction.HORIZONTAL, Direction.VERTICAL, Direction.DIAGONAL_DOWN, Direction.DIAGONAL_UP)
        }

        // Seeded selection per level number to guarantee consistent placement
        val rng = java.util.Random(levelNum * 997L + 13L)
        val shuffledDirs = candidateDirections.shuffled(rng)

        var chosenDirection = Direction.HORIZONTAL
        var chosenStartRow = 0
        var chosenStartCol = 0
        var placed = false

        for (dir in shuffledDirs) {
            val minStartRow = when {
                dir.dRow < 0 -> (word.length - 1)
                else -> 0
            }
            val maxStartRow = when {
                dir.dRow > 0 -> gridRows - word.length
                else -> gridRows - 1
            }

            val minStartCol = when {
                dir.dCol < 0 -> (word.length - 1)
                else -> 0
            }
            val maxStartCol = when {
                dir.dCol > 0 -> gridCols - word.length
                else -> gridCols - 1
            }

            if (maxStartRow >= minStartRow && maxStartCol >= minStartCol) {
                val rowRange = maxStartRow - minStartRow + 1
                val colRange = maxStartCol - minStartCol + 1
                chosenStartRow = minStartRow + rng.nextInt(rowRange)
                chosenStartCol = minStartCol + rng.nextInt(colRange)
                chosenDirection = dir
                placed = true
                break
            }
        }

        if (!placed) {
            chosenDirection = Direction.HORIZONTAL
            chosenStartRow = (gridRows / 2).coerceAtMost(gridRows - 1)
            chosenStartCol = 0
        }

        return LevelData(
            levelNumber = levelNum,
            word = word,
            icon = def.icon,
            hintText = def.hintText,
            gridRows = gridRows,
            gridCols = gridCols,
            startRow = chosenStartRow,
            startCol = chosenStartCol,
            direction = chosenDirection,
            illustrationType = def.illustrationType
        )
    }

    val ALL_100_LEVELS: List<LevelData> by lazy {
        (1..100).map { createLevel(it) }
    }
}
